ha/fault_tolerance/rest_api.py

```diff
@@ -0,0 +1,256 @@
+            if it is false then caller should handle the signals.
+            and needs to call CcRestApi.stop() before application terminates
+        """
+
+        CcRestApi._handle_signals = handle_signals
+
+        # Create Rest Application object and set middleware coroutine to it.
+        # A middleware is a coroutine that can modify either the request or response.
+        CcRestApi._app = web.Application(middlewares=[CcRestApi.rest_middleware])
+
+        # TODO: CORTX-30932 Add routs using view classes and remove below placeholder
+        # Note: Adding route '/' just to test
+        CcRestApi._app.router.add_get('/', CcRestApi.handler_check_alive)
+
+        CcRestApi._app.on_startup.append(CcRestApi._on_startup)
+        CcRestApi._app.on_shutdown.append(CcRestApi._on_shutdown)
+
+    @classmethod
+    async def check_for_unsupported_endpoint(cls, request):
+        """
+        Check unsupported endpoints.
+        Note: this function is expected to be call from middleware.
+        Args:
+            request (web.Request): Request object
+        """
+        # TODO: CORTX-30931 implement
+        pass
+
+    @staticmethod
+    def json_response(resp_obj, status=200):
+        """
+        Returns the the provided response object to client.
+        Args:
+            resp_obj (web.Response): Response object that to be sent to client.
+            status (int, optional): HTTP status code. Defaults to 200.
+        """
+        # TODO: CORTX-30931 Implement
+        pass
+
+    @staticmethod
+    def error_response(err: Exception, **kwargs):
+        """
+        Create response object from input exception object.
+        Args:
+            err (Exception): exception object for that error response sent to be client.
+        """
+        # TODO: CORTX-30931 Implement
+        pass
+
+    @staticmethod
+    @web.middleware
+    async def rest_middleware(request, handler):
+        """
+        The 'middleware' coroutine which get called when Request is received to the server.
+        Args:
+            request (web.Request): Request object
+            handler (handler): handler coroutine
+        Returns:
+            Json object: return json object that to be sent to client, ref: CcRestApi.json_response
+        """
+        Log.debug(f"Rest middleware is called: request = {request} handler = {handler}")
+        if CcRestApi.__is_shutting_down:
+            # TODO : CORTX-30931 return json proper response with HTTP status = 503
+            return CcRestApi.json_response("CC is shutting down", status=503)
+        try:
+            request_id = int(time.time())
+
+            try:
+                await CcRestApi.check_for_unsupported_endpoint(request)
+            except Exception as e:
+                Log.warn(f"Exception: {e}")
+                # TODO:
+                CcRestApi.error_response(Exception("Invalid endpoint."))
+
+            resp = await handler(request)
```

Comment:
Here handler coroutine is handler_check_alive?

Reply:
test method 'handler_check_alive' to be removed after the next dependent task.
added as a place holder.
but all requests will get called through the handler(request) and hence here the method  'handler_check_alive' will also get called.

Reply:
Ok..